cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -776,6 +817,161 @@ protected void doGetSchedules(HttpResponder responder, ApplicationId application
+   * The data is an array of JSON objects
+   * where each object must contain the following three elements: appId, programType, and programId
+   * (flow name, service name, etc.).
+   * <p>
+   * Example input:
+   * <pre><code>
+   * [{"appId": "App1", "programType": "Workflow", "programId": "WF1"},
+   * {"appId": "App1", "programType": "Workflow", "programId": "WF2"}]
+   * </code></pre>
+   * </p><p>
+   * The response will be an array of JsonObjects each of which will contain the three input parameters
+   * as well as a "schedules" field, which is a list of {@link ScheduledRuntime} object.
+   * </p><p>
+   * If an error occurs in the input (for the example above, App1 does not exist), then all JsonObjects for which the
+   * parameters have a valid status will have the status field but all JsonObjects for which the parameters do not have
+   * a valid status will have an error message and statusCode.
+   */
+  @POST
+  @Path("/nextruntime")
+  public void batchNextRunTimes(FullHttpRequest request,
+                                HttpResponder responder,
+                                @PathParam("namespace-id") String namespaceId) throws Exception {
+    List<BatchProgram> batchPrograms = validateAndGetBatchInput(request, BATCH_PROGRAMS_TYPE);
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(batchRunTimes(namespaceId, batchPrograms, false)));
+  }
+
+  /**
+   * Fetches scheduled run times for a set of programs.
+   *
+   * @param namespace namespace of the programs
+   * @param programs the list of programs to fetch scheduled run times
+   * @param previous {@code true} to get the previous scheduled times; {@code false} to get the next scheduled times
+   * @return a list of {@link BatchProgramSchedule} containing the result
+   * @throws SchedulerException if failed to fetch schedules
+   */
+  private List<BatchProgramSchedule> batchRunTimes(String namespace, Collection<? extends BatchProgram> programs,
+                                                   boolean previous) throws Exception {
+    List<ProgramId> programIds = programs.stream()
+      .map(p -> new ProgramId(namespace, p.getAppId(), p.getProgramType(), p.getProgramId()))
+      .collect(Collectors.toList());
+
+    Set<ApplicationId> appIds = programIds.stream().map(ProgramId::getParent).collect(Collectors.toSet());
+    Map<ApplicationId, ApplicationSpecification> appSpecs = store.getApplications(appIds);
+
+    List<BatchProgramSchedule> result = new ArrayList<>();
+    for (ProgramId programId : programIds) {
+      ApplicationSpecification spec = appSpecs.get(programId.getParent());
+      if (spec == null) {
+        result.add(new BatchProgramSchedule(programId, HttpResponseStatus.NOT_FOUND.code(),
+                                            new NotFoundException(programId.getParent()).getMessage(), null));
+        continue;
+      }
+      try {
+        Store.ensureProgramExists(programId, spec);
```

Comment:
not from this PR, but it's kind of misleading to have this method in Store, as it reads like this would be performing some I/O.

Reply:
Agree. it is a static method on the Store interface.